packages/web3-core/src/web3_context.ts

```diff
@@ -1,39 +1,147 @@
 import { Web3APISpec } from 'web3-common';
-import { Web3ConfigOptions, SupportedProviders } from './types';
-import { Web3Config } from './web3_config';
+import { SupportedProviders, Web3ConfigOptions } from './types';
+import { isSupportedProvider } from './utils';
+import { Web3Config, Web3ConfigEvent } from './web3_config';
 import { Web3RequestManager } from './web3_request_manager';
 import { Web3SubscriptionConstructor } from './web3_subscriptions';
 import { Web3SubscriptionManager } from './web3_subscription_manager';
 
+// To avoid circular dependencies, we need to export type from here.
+export type Web3ContextObject<
+	API extends Web3APISpec = any,
+	RegisteredSubs extends {
+		[key: string]: Web3SubscriptionConstructor<API>;
+	} = any,
+> = {
+	config: Web3ConfigOptions;
+	provider: SupportedProviders<API> | string;
+	requestManager: Web3RequestManager<API>;
+	subscriptionManager?: Web3SubscriptionManager<API, RegisteredSubs> | undefined;
+	registeredSubscriptions?: RegisteredSubs;
+	providers: typeof Web3RequestManager.providers;
+};
+
+export type Web3ContextConstructor<
+	// eslint-disable-next-line no-use-before-define
+	T extends Web3Context<any>,
+	T2 extends unknown[],
+> = new (...args: [...extras: T2, context: Web3ContextObject]) => T;
+
+// To avoid circular dependencies, we need to export type from here.
+export type Web3ContextFactory<
+	// eslint-disable-next-line no-use-before-define
+	T extends Web3Context<any>,
+	T2 extends unknown[],
+> = Web3ContextConstructor<T, T2> & {
+	fromContextObject(this: Web3ContextConstructor<T, T2>, contextObject: Web3ContextObject): T;
+};
+
 export class Web3Context<
 	API extends Web3APISpec,
 	RegisteredSubs extends {
 		[key: string]: Web3SubscriptionConstructor<API>;
-	} = {},
+	} = any,
 > extends Web3Config {
 	public static readonly providers = Web3RequestManager.providers;
 	public static givenProvider?: SupportedProviders<never>;
 	public readonly providers = Web3RequestManager.providers;
 
-	public readonly requestManager: Web3RequestManager<API>;
-	public readonly subscriptionManager?: Web3SubscriptionManager<API, RegisteredSubs>;
+	public requestManager: Web3RequestManager<API>;
+	public subscriptionManager?: Web3SubscriptionManager<API, RegisteredSubs>;
 
 	public constructor(
-		provider: SupportedProviders<API> | string,
-		options?: Partial<Web3ConfigOptions>,
-		registeredSubscriptions?: RegisteredSubs,
+		providerOrContext:
+			| SupportedProviders<API>
+			| string
+			| Partial<Web3ContextObject<API, RegisteredSubs>>,
 	) {
 		super();
-		this.requestManager = new Web3RequestManager<API>(provider);
 
-		if (registeredSubscriptions) {
+		if (
+			typeof providerOrContext === 'string' ||
+			isSupportedProvider(providerOrContext as SupportedProviders<API>)
```

Comment:
Would it make sense to include `string` in `SupportedProviders` so you don't have to manually check?